### PR TITLE
Implement starknet_getNonce

### PR DESCRIPTION
**Pull request by jelilat**
Closes https://github.com/NethermindEth/juno/issues/676

### CODE REVIEW

File: rpc/handlers.go [hunk 1]
```diff
@@ -450,3 +451,40 @@ func (h *Handler) Syncing() (*SyncState, *jsonrpc.Error) {
...
+		err = ErrPendingNotSupported
+	} else if id.Hash != nil {
+		stateReader, stateCloser, err = h.bcReader.StateAtBlockHash(id.Hash)
+	} else {
+		stateReader, stateCloser, err = h.bcReader.StateAtBlockNumber(id.Number)
+	}
+
+	if err != nil {
+		return nil, ErrBlockNotFound
+	}
```
Comment (omerfirmak) [inline rpc/handlers.go:482]: Let's make this a function, it will be reused for sure.

Comment (omerfirmak) [reply rpc/handlers.go]: like; https://github.com/NethermindEth/juno/pull/723/files#diff-1f724c5e433be46bfd8dc0962d1796f0dce2a5e22253fe82f69ab6a1eb8df278R456-R467

File: rpc/handlers.go [hunk 2]
```diff
@@ -450,3 +453,41 @@ func (h *Handler) Syncing() (*SyncState, *jsonrpc.Error) {
 		},
 	}, nil
 }
+
+func (h *Handler) readState(id *BlockID) (core.StateReader, func() error, error) {
```
Comment (omerfirmak) [inline rpc/handlers.go:457]: reading from the state doesnt really happen here
how about something like `stateByBlockID`, to bring it inline with `blockByID`, `blockHeaderByID` etc?

Comment (jelilat) [reply rpc/handlers.go:457]: Yeah. Just saw your naming in `stateByBlockID`. Will fix

Comment (jelilat) [reply rpc/handlers.go]: Done

File: rpc/handlers.go [hunk 3]
```diff
@@ -450,3 +453,41 @@ func (h *Handler) Syncing() (*SyncState, *jsonrpc.Error) {
...
+//
+// It follows the specification defined here:
+// https://github.com/starkware-libs/starknet-specs/blob/a789ccc3432c57777beceaa53a34a7ae2f25fda0/api/starknet_api_openrpc.json#L633
+func (h *Handler) Nonce(id *BlockID, address *felt.Felt) (*felt.Felt, *jsonrpc.Error) {
+	stateReader, stateCloser, err := h.readState(id)
+	if err != nil {
+		return nil, ErrBlockNotFound
+	}
+
+	defer func() {
```
[the remark this answers is not in the document]
Comment (jelilat) [reply rpc/handlers.go]: Done

**Comment by omerfirmak** [comment]
Tests are failing due to the nil closer returned by mocks.

File: rpc/handlers_test.go [hunk 4]
```diff
@@ -957,3 +957,77 @@ func TestSyncing(t *testing.T) {
...
+
+		nonce, rpcErr := handler.Nonce(&rpc.BlockID{Number: 0}, &felt.Zero)
+		require.Nil(t, nonce)
+		assert.Equal(t, rpc.ErrBlockNotFound, rpcErr)
+	})
+
+	mockState := mocks.NewMockStateHistoryReader(mockCtrl)
+
+	t.Run("non-existent contract", func(t *testing.T) {
+		mockReader.EXPECT().HeadState().Return(mockState, nil, nil)
```
Comment (omerfirmak) [inline rpc/handlers_test.go:996]: missed one here

**Review by omerfirmak — APPROVED** [review]
Very cool, now let's repeat the same thing for `starknet_getStorageAt` and `starknet_getClassHashAt`

**Pull request merged**